### PR TITLE
RFC: Base calling filter for in situ sequencing

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Overview
As discussed in #1263, this filter component performs base calling for in situ sequencing. The filter determines which base (if any) is represented in each X, Y location in a (round, z plane) slice (i.e., in a (channel, x, y) image). Each pixel value in the resulting image is the base quality score for a given base call.

Currently, the base quality score is calculated as the intensity of the called base divided by the L2 norm of the intensities in all 4 channels (this is how it is done in the original BaristaSeq paper). However, I think it could be useful to add the capability for the user to provide their own custom functions. However, this will likely present challenges for provenance logging. Thoughts, @ambrosejcarr and @dganguli?

# Usage
```python
import numpy as np

from starfish import ImageStack
from starfish.image import Filter

# Make the test image
test = np.ones((1, 4, 1, 2, 2), dtype='uint16') * 0.5

x = [0, 0, 1, 1]
y = [0, 1, 0, 1]

for i in range(4):
    test[0, i, 0, x[i], y[i]] = 1
test[0, 0, 0, 0, 0] = 0.75

# Make the ImageStack
test_stack = ImageStack.from_numpy(test)

# Filter
bc = Filter.CallBases(intensity_threshold=0.8)
base_call = bc.run(test_stack, in_place=False)

base_call.xarray.values
```

In the example above, the test image has intensities as shown below. The field over view is 2 pixels x 2 pixels. Base for channel 0 is "on" in the upper left corner (intensity=0.75), the base for channel 1 is "on" in the upper right hand  corner (intensity=1), the base for channel 2 is "on" in lower left-hand corner (intensity=1), and the base for channel 3 is "on" in the lower right hand corner (intensity=1). The background intensity is uniformly 0.5 for all channels.
```python
array([[[[[0.75, 0.5 ],
          [0.5 , 0.5 ]]],

        [[[0.5 , 1.  ],
          [0.5 , 0.5 ]]],

        [[[0.5 , 0.5 ],
          [1.  , 0.5 ]]],

        [[[0.5 , 0.5 ],
          [0.5 , 1.  ]]]]], dtype=float32)
```
After filtering, we obtain the following. Note that the base 0 intensity in the upper left hand corner was 0.75, so no base was called due to the intensity threshold we set for the filter (`intensity_threshold=0.8`). The base quality for all other bases is `1 / (1^2 + 0.5^2 + 0.5^2 + 0.5^2)^0.5 = 0.755929`.
```python
array([[[[[0.      , 0.      ],
          [0.      , 0.      ]]],

        [[[0.      , 0.755929],
          [0.      , 0.      ]]],

        [[[0.      , 0.      ],
          [0.755929, 0.      ]]],

        [[[0.      , 0.      ],
          [0.      , 0.755929]]]]], dtype=float32)
```
# Testing
I added tests for the base calling and filtering in `starfish/core/image/_filter/test/test_call_bases.py`

### CODE REVIEW

File: starfish/core/image/_filter/call_bases.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            base_qualities < quality_threshold, 0, base_qualities
+        )
+
+        # Threshold the intensity values
+        base_qualities_filtered = xr.where(
+            max_values < intensity_threshold, 0, base_qualities_filtered
+        )
+
+        # Put the base calls in place
+        base_calls = xr.full_like(other=image, fill_value=0)
```
Comment (ambrosejcarr) [inline starfish/core/image/_filter/call_bases.py]: Oh I didn't know about `xr.full_like`. This is great!

File: starfish/core/image/_filter/test/test_call_bases.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                [[0, 0], [0, 0]]], dtype='float32'
+        )
+
+    bases_xr = xr.DataArray(base_calls, dims=(Axes.CH.value, Axes.X.value, Axes.Y.value))
+
+    return bases_xr
+
+
+@pytest.mark.parametrize("int_thresh, qual_thresh", [(0.8, 0), (0, 0.75), (0.8, 0.75)])
+def test_call_bases(int_thresh: float, qual_thresh: float):
```
Comment (ambrosejcarr) [inline starfish/core/image/_filter/test/test_call_bases.py]: Would it be hard to move your example from the PR introduction into the test here? I'm a bit nervous about not having a test that directly consumes the IntensityTable.

Comment (kevinyamauchi) [reply starfish/core/image/_filter/test/test_call_bases.py]: No problem. I can add that as an additional test. I like having an isolated test that tests just the base calling method, so I will also keep the tests currently there.

File: starfish/core/image/_filter/call_bases.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    Parameters
+    ----------
+    intensity_threshold : float
+        Minimum intensity a pixel must have to be called a base.
+        Set to zero for no thresholding.
+    quality_threshold : float
+        Minimum base quality score a pixel must have to be called a base.
+        Set to zero for no thresholding.
+    clip_method : Union[str, Clip]
```
[the remark this answers is not in the document]
Comment (kevinyamauchi) [reply starfish/core/image/_filter/call_bases.py]: Sounds good!

File: starfish/core/image/_filter/test/test_call_bases.py (hunk 2 above)
Comment (ambrosejcarr) [reply starfish/core/image/_filter/test/test_call_bases.py]: Oh yes, definitely. Don't remove these tests. :-)

**Comment by kevinyamauchi** [comment]
@ambrosejcarr , thanks for the comments! I've addressed them. Would you mind re-reviewing?

**Pull request merged**